reps/2022-03-09-shuffle.md

```diff
@@ -0,0 +1,110 @@
+More details on push-based shuffle can be found in the [Magnet paper](https://dl.acm.org/doi/10.14778/3415478.3415558), which implemented the algorithm as a part of Spark.
+
+In the current Datasets shuffle, the metadata overhead for `ObjectRefs` in Ray becomes a bottleneck at around 10TB or more.
+The Exoshuffle work reduces the total number of `ObjectRefs` and showed that it is possible to run sort at 100TB or more data scales on Ray.
+To incorporate this work, we will:
+1. Unify Datasets shuffle-based primitives on the same internal shuffle implementation.
+2. Benchmark Datasets shuffle-based primitives.
+3. Implement a push-based shuffle in Datasets.
+
+### Ray core
+
+Although push-based shuffle reduces the amount of system metadata, it is not enough to scale to petabyte-size data.
+Thus, we also propose a number of Ray core improvements to reduce the total amount of metadata needed during shuffle operations.
+These improvements center around reducing the per-`ObjectRef` metadata needed.
+Currently, each task requires about 5KB of metadata and each object requires about 2KB of metadata at the driver.
+We plan to reduce these by:
+
+1. "Collapsing" `ObjectRef` metadata for objects returned by the same task. All shuffle-on-Ray implementations rely on tasks with multiple return values. Currently, the metadata for these objects is stored separately, but we can combine these to amortize the metadata cost.
+2. "Collapsing" task metadata for tasks submitted in the same stage. This is analogous to the above, but for tasks that are "similar".
+3. Optimizing per-task and per-object metadata. Many of the metadata fields are not actually set or often have the same values. We can potentially save memory by not including these fields.
+
+#### Fault tolerance and scheduling support
+
+The push-based shuffle implemented in Exoshuffle currently requires precise placement of each task to reduce data movement across the cluster.
+Currently, this is implemented using node-specific resource requirements.
+However, this will hang the job if one of those nodes fails.
+
+To ensure fault tolerance, we plan to implement soft scheduling constraints to allow tasks to execute even if their original node fails.
+In the future, we can improve this further by providing a higher-level API to express scheduling constraints for groups of tasks.
+
+## Compatibility, Deprecation, and Migration Plan
+
+This proposal will not include any API changes other than a way to choose the shuffle implementation in Datasets.
+
+## Test Plan and Acceptance Criteria
+The proposal should discuss how the change will be tested **before** it can be merged or enabled. It should also include other acceptance criteria including documentation and examples. 
+
+We plan to benchmark Datasets using the following workloads:
+1. Shuffling data loader for ML ingest.
+2. Sorting.
+3. Groupby, using the [h2oai benchmark](https://h2oai.github.io/db-benchmark/).
+
```

Comment:
optional: should we include chaos test as failure is more common running in large scale?

Reply:
Good idea, yes I'll add it.